core/src/main/java/io/grpc/ServiceProvider.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2017, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc;
+
+/**
+ * An interface for classes that are loaded via {@link java.util.ServiceLoader}. This is not meant
+ * to be implemented directly, but should be used to modify an {@code abstract class}
+ * or {@code interface}.
+ * The {@link ServiceProviders} utility class should be used to create concrete implementations
+ * of the class.
+ */
+interface ServiceProvider {
+  /**
+   * Whether this provider is available for use, taking the current environment into consideration.
+   * If {@code false}, no other methods are safe to be called.
+   */
+  boolean isAvailable();
```

Comment:
I'd really hope these would be `protected`. That does require making this an abstract class though. I'm not sure how the docs work look in either case...

Reply:
Not 100% sure what you're referring to, but I tried an experiment where I created a `public abstract class NettyFoo extends InternalServiceProvider` and a `public class NettyFooImpl extends NettyFoo`. In both cases, `isAvailable` and `priority` show up in the output of `./gradlew javadoc`.

Reply:
Check in the output of all/. It will filter the InternalServiceProvider